zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/client/TestDedicatedZkClient.java

```diff
@@ -0,0 +1,154 @@
+      String realmName = ZK_PREFIX + (ZK_START_PORT + i);
+      RAW_ROUTING_DATA.put(realmName, shardingKeyList);
+    }
+
+    // Feed the raw routing data into TrieRoutingData to construct an in-memory representation of routing information
+    _metadataStoreRoutingData = new TrieRoutingData(RAW_ROUTING_DATA);
+  }
+
+  /**
+   * 1. Create a DedicatedZkClient with a non-existing sharding key (for which there is no valid ZK realm)
+   * -> This should fail with an exception
+   * 2. Create a DedicatedZkClient with a valid sharding key
+   * -> This should pass
+   */
+  @Test
+  public void testDedicatedZkClientCreation() {
+    // Create a DedicatedZkClient
+    String invalidShardingKey = "InvalidShardingKey";
+    RealmAwareZkClient.RealmAwareZkClientConfig clientConfig =
+        new RealmAwareZkClient.RealmAwareZkClientConfig();
+
+    // Create a connection config with the invalid sharding key
+    RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig =
+        new RealmAwareZkClient.RealmAwareZkConnectionConfig(invalidShardingKey);
+
+    try {
+      _realmAwareZkClient = DedicatedZkClientFactory.getInstance()
+          .buildZkClient(connectionConfig, clientConfig, _metadataStoreRoutingData);
+      Assert.fail("Should not succeed with an invalid sharding key!");
+    } catch (IllegalArgumentException e) {
+      // Expected
+    }
+
+    // Use a valid sharding key this time around
+    String validShardingKey = ZK_SHARDING_KEY_PREFIX + "_" + 0; // Use TEST_SHARDING_KEY_0
+    connectionConfig = new RealmAwareZkClient.RealmAwareZkConnectionConfig(validShardingKey);
+    _realmAwareZkClient = DedicatedZkClientFactory.getInstance()
+        .buildZkClient(connectionConfig, clientConfig, _metadataStoreRoutingData);
+  }
+
+  /**
+   * Test the persistent create() call against a valid path and an invalid path.
+   * Valid path is one that belongs to the realm designated by the sharding key.
+   * Invalid path is one that does not belong to the realm designated by the sharding key.
+   */
+  @Test(dependsOnMethods = "testDedicatedZkClientCreation")
+  public void testDedicatedZkClientCreatePersistent() {
+    _realmAwareZkClient.setZkSerializer(new ZNRecordSerializer());
+
+    // Create a dummy ZNRecord
+    ZNRecord znRecord = new ZNRecord("DummyRecord");
+    znRecord.setSimpleField("Dummy", "Value");
+
+    // Test writing and reading against the validPath
+    _realmAwareZkClient.createPersistent(TEST_VALID_PATH, true);
+    _realmAwareZkClient.writeData(TEST_VALID_PATH, znRecord);
+    Assert.assertEquals(_realmAwareZkClient.readData(TEST_VALID_PATH), znRecord);
+
+    // Test writing and reading against the invalid path
+    try {
+      _realmAwareZkClient.createPersistent(TEST_INVALID_PATH, true);
+      _realmAwareZkClient.writeData(TEST_INVALID_PATH, znRecord);
+      Assert.fail("Create() should not succeed on an invalid path!");
+    } catch (IllegalArgumentException e) {
+      // Okay - expected
```

Comment:
It seems `IllegalArgumentException` will be thrown for `createPersistent()`, and there is no check for `writeData()`. If we want to test `writeData()` as well, shall we have another validation for `writeData()` to be more accurate?

Reply:
This is a test for createPersistent(). I added `_realmAwareZkClient.writeData(TEST_INVALID_PATH, znRecord);` for parity with the test I did for the valid path.
_realmAwareZkClient.writeData(TEST_INVALID_PATH, znRecord); doesn't make sense anyway because the ZNode should never be created at that path.